lib/llvmopencl/linker.cpp

```diff
@@ -427,13 +436,153 @@ int copyKernelFromBitcode(const char* name, llvm::Module *parallel_bc,
+                                        llvm::Module *OutputBC) {
+
+  ValueToValueMapTy VVM;
+  llvm::Module::global_iterator GI, GE;
+
+  // Copy all the globals from input to output module.
+  DB_PRINT("cloning the global variables:\n");
+  for (GI = ProgramBC->global_begin(), GE = ProgramBC->global_end(); GI != GE;
+       GI++) {
+    DB_PRINT(" %s\n", GI->getName().data());
+    GlobalVariable *GV = new GlobalVariable(
+        *OutputBC, GI->getValueType(), GI->isConstant(), GI->getLinkage(),
+        (Constant *)0, GI->getName(), (GlobalVariable *)0,
+        GI->getThreadLocalMode(), GI->getType()->getAddressSpace());
+    GV->copyAttributesFrom(&*GI);
+    // for program scope vars, change linkage to external
+    if (isProgramScopeVariable(*GV)) {
+      GV->setDSOLocal(false);
+      GV->setLinkage(GlobalValue::LinkageTypes::ExternalLinkage);
+    }
+    VVM[&*GI] = GV;
+  }
+
+  // add initializers to global vars, copy aliases & MD
+  std::string log;
+  shared_copy(OutputBC, ProgramBC, log, VVM);
+
+  // change program-scope variables in Input module to references
+  std::list<GlobalVariable *> GVarsToErase;
+  std::list<GlobalVariable *> ProgramGVars;
+  for (GI = ProgramBC->global_begin(), GE = ProgramBC->global_end(); GI != GE;
+       GI++) {
+    ProgramGVars.push_back(&*GI);
+  }
+
+  // we can't use iteration over ProgramBC->global_begin/end here, because we're
+  // adding new GVars into the Module during iteration
+  for (auto GI : ProgramGVars) {
+    DB_PRINT(" %s\n", GI->getName().data());
+    if (isProgramScopeVariable(*GI)) {
+      std::string OrigName = GI->getName().str();
+      GI->setName(Twine(OrigName, "__old"));
+      GlobalVariable *NewGV = new GlobalVariable(
+          *ProgramBC, GI->getValueType(), GI->isConstant(),
+          GlobalValue::LinkageTypes::ExternalLinkage,
+          (Constant *)nullptr, // initializer
+          OrigName,
+          (GlobalVariable *)nullptr, // insert-before
+          GI->getThreadLocalMode(), GI->getType()->getAddressSpace());
+
+      NewGV->copyAttributesFrom(GI);
+      NewGV->setDSOLocal(false);
+      GI->replaceAllUsesWith(NewGV);
+      GVarsToErase.push_back(GI);
+    }
+  }
+  for (auto GV : GVarsToErase) {
+    GV->eraseFromParent();
+  }
+
+#if 0
+  // mark SPIRV Kernels as dso_local + external
+  llvm::Module::iterator MI, ME;
+  for(MI = Program->begin(), ME = Program->end() ; MI != ME; ++MI) {
+    Function *F = &*MI;
+    if (F->isDeclaration())
+      continue;
+    if (!F->hasName()) {
+      std::string temp = std::to_string(UniDist(*Mersenne));
+      F->setName(Twine("anonymous_func__",
+                       StringRef(temp)));
+    }
+
+    std::string FName = F->getName().str();
+    if (F->getCallingConv() == llvm::CallingConv::SPIR_KERNEL) {
+      F->setDSOLocal(true);
+      F->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+    }
+  }
+#endif
+
+  // copy functions to the output.bc, then replace them with references
```

Comment:
This parts needs better documentation: What is done here and _why_. Why "move non kernels"?

Reply:
fixed